### ISSUE TITLE
Switch to Vinyl

### ISSUE THREAD

**Issue by dominicbarnes** [makojs/tree#9]
I want to seriously consider replacing the `File` internals here with [vinyl](https://www.npmjs.com/package/vinyl). This issue will be my brainstorming/scratchpad for this transition.

Let's start with the benefits:

 - Going 3rd-party means less to maintain in core
 - Gulp has been using this for a long time, and there are a lot of really solid design decisions they've made (as they continue to press it forward, I can ride their coat-tails)
 - The internals of `path` and `history`  is just more flexible and intuitive than what I have
 - The addition of `base` being included (along with `cwd`) is also more robust that what I have

To successfully switch, there are a few things I'll need to rework/consider:

First, I've been relying on `file.path` as an "immutable" ID for files in the tree, which doesn't always make sense, especially if:

 - you want to indicate the file needs to be output elsewhere (a separate `file.output` is really awkward)
 - files go through several steps of change throughout the build (eg: transpile and change extension)
 - you are adding build-only files, like `*.map` and `shared.js`

I think the better long-term solution here is to generate UUIDs for each file and use other mechanisms for querying by path. This could have other implications for plugins, as finding other files could become less straightforward. I'm hoping that is not the case, as long as things like `File#getDependencies()` still continues to work as-is.

Second, I'll need to make sure that behavior like `File#dirty()` can continue working as well. In fact, I'll need to be careful about all the props I've started to rely on internally.

Third, I don't (yet) know of a way to serialize a vinyl object so it can be saved using mako-cache. I'm sure there's another module for just this case, I just need to do more research.

Fourth, by adopting vinyl, I'm not sure how well mako will interoperate with other vinyl-* plugins. If those plugins only really work within Gulp, I could be setting myself up to just create confusion for mako plugin devs. (where they assume things are more Gulp-like than they really are)

Fifth, vinyl does not use strings for `file.contents`, the closest thing is buffers, which will be an extra step that plugin authors will need to deal with. It's probably for the best, since memory-usage for buffers is better than with strings, plus vinyl complains pretty loudly when you do the wrong thing.

I'll start a branch very soon to start playing around with this concept. It looks like it'll be a pretty big disruption to mako, but I think it'll be for the best, especially so early on in the project.

**Comment by darsain** [comment]
> I'm not sure how well mako will interoperate with other vinyl-* plugins. If those plugins only really work within Gulp, I could be setting myself up to just create confusion for mako plugin devs. (where they assume things are more Gulp-like than they really are)

tl;dr: adopting Vinyl will mostly help mako internals and API. You won't be able to use existing gulp plugins in mako internals, unless you want to rewrite everything to use streams instead of events, which I don't think will be the case :) The interoperability with gulp ecosystem it'll add is there, but I'd look at it as a secondary benefit.

---

### The gulp/vinyl ecosystem

There aren't really *vinyl plugins* if we want to be strict about terminology. There are *vinyl adapters* which consume files from various sources, and output object streams of vinyl files. For example, [vinyl-fs](https://github.com/gulpjs/vinyl-fs) adapter reads files from filesystem and creates a readable stream of vinyl objects. There are also smaller vinyl modules for filesystem like [vinyl-file](https://github.com/sindresorhus/vinyl-file) and [vinyl-read](https://github.com/SamVerschueren/vinyl-read), but again, all they are is a means to generate virtual vinyl files.

What works with these objects are called *gulp plugins*, which are just transform streams that can pipe to readable vinyl streams and do something with them. The usual gulp task looks like this:

```js
const gulp = require('gulp');
const concat = require('gulp-concat');

gulp.task('concat', () => {
	return gulp.src('*.js')        // readable vinyl object stream of all *.js files
		.pipe(concat('all.js'))    // transform stream that concatenates all *.js into all.js
		.pipe(gulp.dest('build')); // writable stream that saves all it reads into build folder
});
```

The gulp ecosystem is than just a collection of transform streams that mutate vinyl objects, and can be easily used outside of gulp.

The same function completely detached from gulp, and usable in any other application, would look like this:

```js
const vfs = require('vinyl-fs');
const concat = require('gulp-concat');

function concatenate(globs, name, to, done) {
	vfs.src(globs)
		.pipe(concat(name))
		.pipe(vfs.dest(to))
		.on('finish', done);
}

concatenate('*.js', 'all.js', 'build', cb);
```

### Interoperability with gulp

Here is my [mako vinyl adapter](https://gist.github.com/darsain/5910d58e1d5556b5b384aa1a9f94c008) that allows me to use mako with gulp (all it does is wrap `mako#use()` and makes `mako#build()` return a vinyl stream instead of a promise).

If mako would use Vinyl:

1. I could remove the `VinylMakoStream#vinyl()` method which transforms mako file to vinyl file.
2. The whole adapter could than be implemented as an alternative `mako#buildStream()` method.
3. If `mako#build(...entries)` would accept vinyl objects, `mako#buildStream()` could be made a transform stream. That would remove the need to create a mako plugin for every pre/post-processor out there, and allow us to just use the already existing gulp plugins.

For example, this would use gulp-stylus to generate CSS and pass it to mako:

```js
// imagine nice wall of requires here

function buildStyles() {
	let makoTransformStream = mako()
		.use(stat(['css', ...asets]))
		.use(text(['css']))
		.use(buffer(assets))
		.use(css())
		.buildStream();

	return vfs.src('./index.styl')
		.pipe(stylus())            // gulp-stylus plugin
		.pipe(makoTransformStream)
		// other gulp plugins here
		.pipe(vfs.dest('build'));
}
```

### Vinyl in mako core

The benefit of Vinyl is mainly all the other things you've mentioned, like more thought out spec, property handling, history, nicer API, etc. Maybe also familiarity with developers experienced with gulp.

Also, you can add your own properties like `dirty` to vinyl object. Vinyl thinks about it in it's `Vinyl#clone()` method. The only standardized custom property (in a sense that a lot of gulp plugins consume/work with it) I know of right now is `sourceMap`, so if a file gets a sourcemap, it should go there. Apart of that, the file object is yours to pollute.

Of course, if you need custom methods, extending vinyl into new constructor is necessary.

---

Hope this made the whole `mako<>vinyl<>gulp` thing clearer.

**Comment by dominicbarnes** [comment]
Thanks for the great explanation @darsain! I'll be working on implementing Vinyl here in the tree. (most likely by extending the constructor) I've got a lot of other hurdles to overcome in mako core before I'm done, so it could be a while. I'll continue opening PRs to give you (and anyone else who cares) greater visibility into what I'm doing. :)

**Comment by darsain** [comment]
I don't remember last time I've seen some plugin using map-stream. Usually everything uses [through2](https://www.npmjs.com/package/through2). Dunno if that has the same issue?

**Comment by matthewmueller** [comment]
sweet, i *think* through2 should be fine

**Comment by matthewmueller** [comment]
Starting to think about this stuff again :-)

> First, I've been relying on file.path as an "immutable" ID for files in the tree, which doesn't always make sense, especially if:

I really like this idea, what about using an MD5 hash as the ID? By having that assumption, you might be able to do some cool optimizations internally regarding file changes.

**Comment by dominicbarnes** [comment]
@matthewmueller I'm actually just thinking about generating UUIDs for each file, as I need something that won't change throughout the build. MD5 hashes can still be incorporated via plugins though. :)

**Comment by matthewmueller** [comment]
yah okay that makes sense to me, i forgot that MD5 can change throughout the life of the build :-D

**Comment by matthewmueller** [comment]
One question I have about this is... where does `root` fit into all of this? Currently we're passing in root to each plugin that needs it:

```js
mako()
  .use(output('js', { root: __dirname }))
  .build(entries)
```

But it seems like with this change, the output plugin will go away and relying on `process.cwd()` isn't robust enough. Personally, I'd like a place where I can set the root and have access to it from all the plugins, since I'll have access to the `file.base` in each of the plugins. 

Perhaps something like this:

```js
let config = {
  root: __dirname
}

Mako(config)
  .use(JS('js'))
  .build(entries)
```

Then internally: 

```
let tree = new Tree(root)
tree.addDependency('index.js') // adds the root as the base and index.js as the path
```

**Comment by dominicbarnes** [comment]
@matthewmueller I'm still working out the details, but I believe `file.base` should be enough to act as "root".

You will be able to pass `{ base: '/some/dir', path: '/some/dir/some/file' }` as the entry files, rather than just plain strings. (if you use plain strings, base is assumed to be cwd) Through the process of adding new files, the base will be replicated as well. That said, plugins will need to respect this behavior, so I'm going to finish my current series of changes before I consider adding "root" to the tree itself.

The main reason I avoided having mako aware of global config was to reduce internal complexity. That being said, having 1 global config like root probably isn't so bad.

Also, the output plugin doesn't go away, it simply changes to use the newer `file.path` mechanisms, and no longer deals with `file.output`.

**Comment by matthewmueller** [comment]
Ah okay. Yah, I think some sort of global config would be pretty convenient. Webpack does a pretty good job with this, they take a config file that's actually just a JS file, so you can do stuff like:

```js
module.exports = {
  sourcemaps: process.env.NODE_ENV === 'production' ? 'hidden' : 'inline'
}

let webpack = Webpack(require('./webpack.js')
```

I think that would make mako pretty flexible, while easy to setup. 

---

Maybe I just should just open a new issue haha, but I think you should go either all live (file.path updates automatically whenever file.type changes) or all recorded (keep old file.path and maintain a file.output key).

I'd actually prefer keeping all these keys around. I found a case where I'd like to be able to map the original file to the built file. If you want to automatically switch between serving a file directly in dev vs. a built file in production, it's nice to maintain that mapping.

**Comment by darsain** [comment]
I still think that using `file.cwd` and/or `file.base` (already present on vinyl file objects) is better than one global root config, since than you can use the same mako instance to build multiple files with different roots. In my head this is a per entry config rather than per mako instance. But I have nothing against having one global root config that everything will default to unless overridden in an entry object. That might be useful. But definitely move away from `option.root` in every plugin :)

---

Also, the `file.contents` being forced to buffer, ... it's kind of unfortunate. It's a good idea, since buffers are faster and more efficient to work with, but practically no module works with them. They all convert to strings, apply some transformations, and than convert back to buffer, effectively creating more overhead than if it'd be a string from the beginning.

Did you decide to just extend vinyl and go the always buffer way, or did you clone the whole thing and removed the content type enforcing?
